Makes the remoteStatus change synchronous for media uploads.

diegoreymendez opened the pull request:
Second attempt at resolving this issue.  The first one was reverted due to a deadlock: https://github.com/wordpress-mobile/WordPress-iOS/pull/12450

## Description:

While trying to resolve #12416, I realized that media being uploaded wasn't properly marked as having `remoteStatus == .pushing`.  Upon debugging this issue I realized the problem was that we are setting this status asyncrhonously here:

https://github.com/wordpress-mobile/WordPress-iOS/blob/534556756a7d11efb3e10000aad6860493bc7d6b/WordPress/Classes/Services/MediaService.m#L245-L252

In this PR I change that block to be executed synchronously, as well as the save operation at the end of the block.

## Testing:

Place breakpoint A [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/b36c55c1259ab67893e7a63b213c98e8970f2a7b/WordPress/Classes/Services/MediaService.m#L219).
Place breakpoint B [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/b36c55c1259ab67893e7a63b213c98e8970f2a7b/WordPress/Classes/Services/MediaService.m#L225)

### Test:

1. Launch the editor and insert an image in a post.
2. Breakpoint A should be hit.
3. Do `po mediaInContext.remoteStatus` and make sure the remote status is `MediaRemoteStatusLocal`.
4. Step once, `po mediaInContext.remoteStatus` and make sure it's now `MediaRemoteStatusPushing`.
5. Tap play, breakpoint B should hit.
6. Do `po mediaInContext.remoteStatus` and make sure the remote status is still `MediaRemoteStatusPushing`. - ie: it was synchronously saved

## Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

diegoreymendez commented on WordPress/Classes/Services/MediaService.m:
```diff
@@ -208,6 +208,19 @@ - (void)uploadMedia:(Media *)media
+    [synchronousContext performBlockAndWait:^{
+        Media *mediaInContext = (Media *)[synchronousContext existingObjectWithID:mediaObjectID error:nil];
+        if (mediaInContext) {
+            mediaInContext.remoteStatus = MediaRemoteStatusPushing;
```
Yup, that's correct.  I'm planning to do that right after this PR, so that the code is more organized and better tested.

diegoreymendez replied:
I'm marking this as resolved as we're in agreement.

diegoreymendez commented on WordPress/Classes/Services/MediaService.m:
```diff
@@ -242,14 +255,6 @@ - (void)uploadMedia:(Media *)media
         return;
     }
 
-    [self.managedObjectContext performBlock:^{
```
The original issue wasn't that though.  The issue I was having is that after calling `uploadMedia()`, a subsequent CD query could still return that media object as still being `.local` instead of `.pushing`, thus not recognizing there was an upload in progress for the media object.

shiki replied:
@diegoreymendez Do you have some reproduction steps for that so I can test if this PR fixes the problem?

diegoreymendez replied:
Basically this is what's causing issues with test case 5 here: https://github.com/wordpress-mobile/WordPress-iOS/pull/12416

The closest thing I can provide to being able to test the issue is to follow the testing steps detailed in this PR against `develop` and once you hit step 6 notice the `remoteStatus` for the post is still `.local`.

You can try: `po MediaCoordinator.shared.isUploadingMedia(for: post)` which will return `false` (this method checks the `remoteStatus` of posts).

diegoreymendez commented on WordPress/Classes/Services/MediaService.m:
```diff
@@ -208,6 +208,19 @@ - (void)uploadMedia:(Media *)media
+        if (mediaInContext) {
+            mediaInContext.remoteStatus = MediaRemoteStatusPushing;
+            mediaInContext.error = nil;
+            [[ContextManager sharedInstance] saveContextAndWait:synchronousContext];
```
I've been discussing about our CoreData Stack with @astralbodies and we feel some of the problems that are surfacing may be because our CD Stack could use a refresh.

In this particular instance, the actual problem I'm seeing is that locking a child context to synchronously save the parent contexts is a sure-recipe for deadlocks.

But from the architectura PoV: I don't think a child context should be the one to decide when a parent context is saved.  This is less of a problem in our async methods for obvious reasons.

I realize what I'm saying isn't in-line with our existing architecture, but this is one case where I feel we need to take a moment to consider addressing the core issue.

I'll think about it a bit more and post my conclusions here.

diegoreymendez replied:
For now I'm following the existing architecture, and letting a synchronous save operation in a child context force the parent contexts to save (asynchronously).  We'll probably review all this at the GM with @astralbodies.

diegoreymendez commented:
Hey folks, apologies for the back and forth.  I introduced some changes that should fix the deadlocks, and this is **ready for review again**.

### An important note:

Using a single `backgroundContext` across threads and unrelated service calls isn't good:

From: https://developer.apple.com/documentation/coredata/nsmanagedobjectcontext

> Core Data uses thread (or serialized queue) confinement to protect managed objects and managed object contexts (see Core Data Programming Guide). A consequence of this is that a context assumes the default owner is the thread or queue that allocated it—this is determined by the thread that calls its init method. You should not, therefore, initialize a context on one thread then pass it to a different thread.

One of the changes I did to address that in the particular issue I'm dealing with, is to instantiate a derived context for my use, and discard it after I'm done: https://github.com/wordpress-mobile/WordPress-iOS/blob/31321dcd193379f57fee401d20904d4b0f04080e/WordPress/Classes/Services/MediaCoordinator.swift#L258

leandroalonso commented:
@diegoreymendez this seems to have fixed all the freezes we were seeing. 🎉 

But apparently it added a regression (which I can't reproduce in `develop`):

Go offline
1. Open the app
2. Add a post with 2 or more images
3. Save as draft
4. Close the app
Go online
1. Open the app
2. Check that all images were uploaded but the path of the images will still point to the local path

```
<img src="file:///Users/leandroalonso/Library/Developer/CoreSimulator/Devices/E690FA1D-AE36-4267-905D-8F6E71F4FA31/data/Containers/Data/Application/E5A8E427-9123-43E7-BE4E-AD701EF5F75B/Library/Caches/Media/thumbnail-p138-1792x1792.jpeg" class="size-full" data-wp_upload_id="x-coredata://5B3134DC-DD18-485C-836C-A68083BE426E/Media/p138"><img src="file:///Users/leandroalonso/Library/Developer/CoreSimulator/Devices/E690FA1D-AE36-4267-905D-8F6E71F4FA31/data/Containers/Data/Application/E5A8E427-9123-43E7-BE4E-AD701EF5F75B/Library/Caches/Media/thumbnail-p139-1792x1792.jpeg" class="size-full" data-wp_upload_id="x-coredata://5B3134DC-DD18-485C-836C-A68083BE426E/Media/p139">
```

The curious thing is that it works every time if you use a single image.

@shiki and @yaelirub, can you please try to reproduce the same behavior?

shiki commented:
I can confirm the regression that @leandroalonso described in https://github.com/wordpress-mobile/WordPress-iOS/pull/12459#issuecomment-528522208.

shiki commented on WordPress/Classes/Utility/ContextManager.m:
```diff
@@ -133,9 +133,32 @@ - (void)saveDerivedContext:(NSManagedObjectContext *)context withCompletionBlock
+            // to follow the existing architecture, we can at least ensure saving the parent is safe, and
+            // that we're not deadlocking the current thread to do that.
+            //
+            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
```
I'm not sure I understand. The way I understood the problem is that, as you described in https://github.com/wordpress-mobile/WordPress-iOS/pull/12459/files#r321340875, we want the data to be persisted right away so that any fetch executing somewhere else would receive the up to date data. But doesn't running `dispatch_async` here prevent the `context.parentContext` to be persisted **right away**? And hence, it's still possible for fetch requests somewhere else to receive stale data?

What was causing the deadlock in here anyway?

diegoreymendez replied:
> But doesn't running dispatch_async here prevent the context.parentContext to be persisted right away?

This is correct, but we only need it to be persisted right away to the main context - saving our derived context (the context the service is using) does just that.  Once the changes hit the main context, they're available across the entire App.

The save to disk (which is what happens when we save the parent context / main context) can be asynchronous (in fact due to our `ContextManager`'s architecture [it's always async no matter what we do](https://github.com/wordpress-mobile/WordPress-iOS/blob/b862ae4ad4f50748bb02b98bdab694f4d197da22/WordPress/Classes/Utility/ContextManager.m#L277-L290)).

As a matter of fact I don't think a derived context should attempt to save to disk at all, or to save parent contexts but that's a deeper discussion beyond the scope of this PR.  This is why I don't feel it'd be right to look for deeper changes in this PR.

> And hence, it's still possible for fetch requests somewhere else to receive stale data?

Not with the code as it is.

This could only happen if we called `uploadMedia()` with a context that doesn't have the main context as it's parent.  But again, picking a context that has the main context as its parent is a deliberate choice and the responsibility of the caller.  This part is important - not any context is the same when calling a service.

diegoreymendez commented:
Down the rabbit hole!  Thanks both, I'll check this.

diegoreymendez commented:
Hey folks, the regression has been fixed.  But the reason behind this regression is actually a nice example why making this service synchronous is a good thing to begin with.

## Some background @leandroalonso 's recent PR:

The PR: 

@leandroalonso 's changes introduced two extra verifications to make sure we're either uploading media, or there's failed media in a post:

https://github.com/wordpress-mobile/WordPress-iOS/blob/0b497dcdb536c7e805d2e6cb6948096d6f8eafa2/WordPress/Classes/Services/PostCoordinator.swift#L62

https://github.com/wordpress-mobile/WordPress-iOS/blob/0b497dcdb536c7e805d2e6cb6948096d6f8eafa2/WordPress/Classes/Services/PostCoordinator.swift#L78

But if you think about it, why would there be failed media in the post, if we just retried uploading all media?

https://github.com/wordpress-mobile/WordPress-iOS/blob/0b497dcdb536c7e805d2e6cb6948096d6f8eafa2/WordPress/Classes/Services/PostCoordinator.swift#L51-L58

The reason is simple - in Leandro's PR, the upload service was correctly setting the remoteStatus of media as `.pushing`, but it was still doing so asynchronously.

https://github.com/wordpress-mobile/WordPress-iOS/blob/521cea4d2465271f6ed9c936297902826551994c/WordPress/Classes/Services/MediaService.m#L245-L252

This meant that his check for `remoteStatus == .failed` iin the IF condition was necessary because the asynchronous block hadn't been executed yet.

## Why did this PR introduce a regression?

As soon as I changed that block to execute synchronously, the IF conditions introduced by @leandroalonso were no longer valid.

Now I need to check if there is media being pushed, because the remote status is being updated *in time*:

https://github.com/wordpress-mobile/WordPress-iOS/blob/7d4570ecba18287b4210d386cf81698ad514796a/WordPress/Classes/Services/PostCoordinator.swift#L68

https://github.com/wordpress-mobile/WordPress-iOS/blob/7d4570ecba18287b4210d386cf81698ad514796a/WordPress/Classes/Services/PostCoordinator.swift#L84

## There's another issue!

But it's 100% beyond the scope of this PR.

We fixed part of the problem with our work, but if you check the IF conditions above (both in @leandroalonso and my version of them) the conditions don't make a lot of sense.

Due to the `remoteStatus` being updated in time, it's much clearer in this PR why the conditions don't make sense though:

```swift
if mediaCoordinator.isUploadingMedia(for: post) || post.hasLocalMediaBeingUploaded() {
```

We're checking twice the same thing.  And indeed we have something that's preventing the first part from returning true even if there *IS* media being uploaded.

This should be tackled separately though, as it's not blocking our offline work, and it's not related specifically to what this PR is trying to fix.

leandroalonso commented:
I'm still able to reproduce the issue consistently:

Being Offline:

1. Create a post and add 2 images
2. Save as Draft
3. Close the app
4. Go online
5. Check that all media are uploaded but the paths are not updated (check the HTML post's content)

At first, I thought it was a problem with my connection, but the same steps seem to work in `develop`.

diegoreymendez commented:
Closing this.  There's some validity in the changes proposed here but we need a much bigger discussion and analysis before moving forward with this now.

Also [the PR that needed these changes](https://github.com/wordpress-mobile/WordPress-iOS/pull/12416) resolved the problem through a workaround.
